### PR TITLE
Add SQ8-FP16 scalar kernels, dispatchers, tests and benchmarks [MOD-15141]

### PR DESCRIPTION
**Pull request by dor-forer**
**Describe the changes in the pull request**

Adds asymmetric SQ8-to-FP16 distance support (scalar path only) for L2, IP, and Cosine, where stored vectors are SQ8-quantized (uint8 + FP32 metadata) and queries are FP16. This is the first step (P1a) of [MOD-15141](https://redislabs.atlassian.net/browse/MOD-15141); SIMD kernels (P1b/P1c) and `QuantPreprocessor` FP16 input support (P1a-2) are intentionally out of scope and tracked as follow-ups.

The scalar kernels use the algebraic identity `dist = ||x||² + ||y||² - 2 * IP(x, y)` with precomputed FP32 sums-of-squares stored alongside both the SQ8 storage and FP16 query, which keeps the quantization-error path consistent between the kernel and the test baseline. Each FP16 query element is widened to FP32 during accumulation; SQ8 metadata (`min`, `delta`, `x_sum`, `x_sum_squares`) and FP16 query metadata (`y_sum`, `y_sum_squares`) are read as FP32.

The new `GetDistFunc<sq8, float, float16>` specialization routes through `IP_SQ8_FP16_GetDistFunc`, `Cosine_SQ8_FP16_GetDistFunc`, and `L2_SQ8_FP16_GetDistFunc`. These dispatchers currently always return the scalar implementation; SIMD chooser slots are reserved for P1b (MOD-15152) and P1c (MOD-15153).

**Which issues this PR fixes**
1. MOD-15141

**Main objects this PR modified**
1. `src/VecSim/spaces/IP/IP.{h,cpp}` — adds `SQ8_FP16_InnerProduct_Impl`, `SQ8_FP16_InnerProduct`, and `SQ8_FP16_Cosine`.
2. `src/VecSim/spaces/L2/L2.{h,cpp}` — adds `SQ8_FP16_L2Sqr` using the algebraic identity over the precomputed sums-of-squares.
3. `src/VecSim/spaces/IP_space.{h,cpp}` — adds `IP_SQ8_FP16_GetDistFunc` and `Cosine_SQ8_FP16_GetDistFunc` dispatchers.
4. `src/VecSim/spaces/L2_space.{h,cpp}` — adds `L2_SQ8_FP16_GetDistFunc` dispatcher.
5. `src/VecSim/spaces/spaces.cpp` — adds the `GetDistFunc<vecsim_types::sq8, float, vecsim_types::float16>` specialization wiring L2/IP/Cosine to the new dispatchers.
6. `tests/utils/tests_utils.h` — adds `populate_sq8_fp16_query` and `SQ8_FP16_NotOptimized_{InnerProduct,Cosine,L2Sqr}` baselines (the L2 baseline mirrors the kernel's algebraic identity to avoid quantization-error drift at high `dim`).
7. `tests/unit/test_spaces.cpp` — adds `SQ8_FP16_NoOptimizationSpacesTest` parameterized over 19 dimensions (including odd and SIMD-boundary residues: 1, 5, 7, 8, 9, 15, 16, 17, 31, 32, 33, 47, 48, 49, 63, 64, 65, 127, 128) for L2/IP/Cosine, plus `SQ8_FP16_EdgeCases` (zero query, constant storage, mixed-sign), `SpacesTest.GetDistFuncSQ8FP16Asymmetric`, and `SpacesTest.GetDistFuncInvalidMetricSQ8ToFP16`.
8. `tests/benchmark/spaces_benchmarks/bm_spaces_sq8_fp16.cpp` (new) + `tests/benchmark/CMakeLists.txt` — registers naive scalar benchmarks for L2/IP/Cosine; SIMD benchmark blocks will be added by P1b/P1c when the chooser symbols become available.

**Test results**
- All 1096 `test_spaces` tests pass locally.
- `bm_spaces_sq8_fp16` builds and runs successfully.

**Follow-ups (separate stories, out of scope here)**
- MOD-15152 (P1b) — x86 SIMD kernels (AVX2 / AVX512 / AVX512_VNNI / SSE4) and benchmark wiring.
- MOD-15153 (P1c) — ARM SIMD kernels (NEON / SVE / SVE2) and benchmark wiring.
- MOD-15141 P1a-2 — `QuantPreprocessor` FP16 input support (no plumbing into the index types yet).

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

[MOD-15141]: https://redislabs.atlassian.net/browse/MOD-15141?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---

> [!NOTE]
> **Medium Risk**
> Touches core distance computation and dispatch for a new datatype combination; primary risk is subtle correctness/UB issues around mixed-type layouts and unaligned metadata reads, mitigated by added `load_unaligned` and broad test coverage.
> 
> **Overview**
> Adds asymmetric **SQ8 storage + FP16 query** support for `IP`, `Cosine`, and `L2` by introducing new scalar kernels (`SQ8_FP16_InnerProduct_Impl`/`SQ8_FP16_InnerProduct`/`SQ8_FP16_Cosine` and `SQ8_FP16_L2Sqr`) that widen FP16 elements to FP32 and use precomputed SQ8/query metadata.
> 
> Wires the new mode into the public dispatch path via `GetDistFunc<sq8, float, float16>` and new `*_SQ8_FP16_GetDistFunc` helpers (currently always returning scalar implementations), and adds `load_unaligned` to safely read trailing FP32 metadata for odd dimensions.
> 
> Extends validation and performance tooling with a new `bm_spaces_sq8_fp16` benchmark target plus unit tests (including odd-dimension unaligned-metadata and edge cases) and new SQ8-FP16 baseline/query-population utilities in `tests_utils.h`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 304450db202fd8f1629099cd141534ab48f3ee9e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/VecSim/utils/alignment.h [hunk 1]
```diff
@@ -8,6 +8,23 @@
...
+// Compilers reliably lower this to a single load on architectures that allow
+// unaligned access; on strict-alignment targets it expands to a safe byte copy.
+template <typename T>
+static inline T load_unaligned(const void *ptr) {
+    static_assert(std::is_trivially_copyable_v<T>,
+                  "load_unaligned requires a trivially-copyable T");
+    T value;
+    std::memcpy(&value, ptr, sizeof(T));
+    return value;
+}
```
Comment (lerman25) [inline src/VecSim/utils/alignment.h:26]: Same alignment concern as [#discussion_r3161710275](https://github.com/RedisAI/VectorSimilarity/pull/942#discussion_r3161710275), but on the existing SQ8_FP32 / SQ8_SQ8 kernels — they read FP32 metadata via a direct reinterpret_cast<const float*> and have the same misalignment for odd dim. Can we either apply load_unaligned there too in this PR, or open a follow-up so we don't end up with two conventions for the same problem in adjacent functions?

Comment (dor-forer) [reply src/VecSim/utils/alignment.h:26]: You're right. 
This is not in the scope of this task - I opened a ticket for this:
https://redislabs.atlassian.net/browse/MOD-15303

**Review by lerman25 — APPROVED** [review]
Good job!
